torch/ao/quantization/_pt2e/qat_utils.py

```diff
@@ -0,0 +1,137 @@
+)
+
+def _conv_bn_pattern(
+    x: torch.Tensor,
+    conv_weight: torch.Tensor,
+    conv_bias: torch.Tensor,
+    bn_weight: torch.Tensor,
+    bn_bias: torch.Tensor,
+    bn_running_mean: torch.Tensor,
+    bn_running_var: torch.Tensor,
+) -> torch.Tensor:
+    x = F.conv2d(x, conv_weight, conv_bias)
+    x = F.batch_norm(x, bn_running_mean, bn_running_var, bn_weight, bn_bias, training=True)
+    return x
+
+def _fused_qat_conv_bn_pattern(
+    x: torch.Tensor,
+    conv_weight: torch.Tensor,
+    conv_bias: torch.Tensor,
+    bn_weight: torch.Tensor,
+    bn_bias: torch.Tensor,
+    bn_running_mean: torch.Tensor,
+    bn_running_var: torch.Tensor,
+) -> torch.Tensor:
+    """
+    Approximated method to fuse conv and bn. It requires only one forward pass.
+    conv_orig = conv / scale_factor where scale_factor = bn.weight / running_std.
+    This is based on `nniqat.ConvBn2d._forward_approximate`.
+    """
+    # TODO: allow setting eps
+    bn_eps = 1e-5
+    running_std = torch.sqrt(bn_running_var + bn_eps)
+    scale_factor = bn_weight / running_std
+    weight_shape = [1] * len(conv_weight.shape)
+    weight_shape[0] = -1
+    bias_shape = [1] * len(conv_weight.shape)
+    bias_shape[1] = -1
+    scaled_weight = conv_weight * scale_factor.reshape(weight_shape)
+    zero_bias = torch.zeros_like(conv_bias, dtype=x.dtype)
+    x = F.conv2d(x, scaled_weight, zero_bias)
+    x = x / scale_factor.reshape(bias_shape)
+    x = x + conv_bias.reshape(bias_shape)
+    x = F.batch_norm(x, bn_running_mean, bn_running_var, bn_weight, bn_bias, training=True, eps=bn_eps)
+    return x
+
+def _get_aten_graph_module(
+    pattern: Callable,
+    example_inputs: Tuple[Any, ...],
+) -> GraphModule:
+    """
+    Convert the pattern to an FX graph with decomposed aten ops.
+    """
+    aten_pattern, _ = torch._dynamo.export(
+        pattern,
+        *copy.deepcopy(example_inputs),
+        aten_graph=True,
+        tracing_mode="real",
+    )
+    aten_pattern.graph.eliminate_dead_code()
+    aten_pattern.recompile()
+    return aten_pattern
+
+def _fuse_conv_bn_qat(m: GraphModule) -> GraphModule:
+    """
+    Given a graph of decomposed aten ops, replace the (conv + bn) pattern with
+    the fused QAT subgraph equivalent. The input graph should already be annotated.
+    The annotations in the original nodes will be preserved in the corresponding
+    nodes in the new subgraph.
+    """
+    m.graph.eliminate_dead_code()
+    m.recompile()
+    example_inputs = _conv_bn_pattern_example_inputs
+    match_pattern = _get_aten_graph_module(_conv_bn_pattern, example_inputs)
+    replacement_pattern = _get_aten_graph_module(_fused_qat_conv_bn_pattern, example_inputs)
+    match_and_replacement = _replace_pattern(m, match_pattern, replacement_pattern)
+    m.recompile()
+
+    # Copy over metadata from original subgraph
+    # This ensures the stack traces and annotations are preserved in the new subgraph
+    for mr in match_and_replacement:
```

Comment:
Also for this if we can have subgraph rewriter support it, it might be great. We did this for TS variant here https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/passes/subgraph_rewrite.cpp#L56, but probably harder to do here since you dont have access to internal nodes of the pattern

Reply:
Added a TODO